src/transformStateWithClones.js

```diff
@@ -7,7 +7,25 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    if (action.type === 'clear') {
+      currentState = {};
+    } else if (action.type === 'addProperties') {
+      currentState = { ...currentState, ...action.extraData };
+    } else if (action.type === 'removeProperties') {
+      currentState = { ...currentState };
+
+      for (const key of action.keysToRemove) {
+        delete currentState[key];
+      }
+    }
+    stateHistory.push(currentState);
```

Comment:
The `stateHistory.push(currentState)` should be moved outside of the `if-else` block to ensure that the state is pushed to the history after processing each action. This aligns with the guideline to push a copy of the object to the array at the end of each loop cycle, but outside of the `switch` block.